cache/sizedlru/sizedlru.go

```diff
@@ -0,0 +1,111 @@
+// Package sizedlru implements an LRU cache that bounds the total space occupied
+// by the items it contains.
+
+package sizedlru
+
+import (
+	"container/list"
+)
+
+type SizedItem interface {
+	Size() int64
+}
+
+// SizedLRU is an LRU cache that will keep its total size below maxSize by evicting
+// items. If maxSize is zero, eviction is disabled.
+// SizedLRU is not thread-safe.
+type SizedLRU struct {
+	// Eviction double-linked list. Most recently accessed elements are at the front.
+	ll *list.List
+	// Map to access the items in O(1) time
+	cache       map[interface{}]*list.Element
+	currentSize int64
+	// SizedLRU will evict items as needed to maintain the total size of the cache
+	// below maxSize.
+	maxSize int64
+}
+
+type Key interface{}
+
+type entry struct {
+	key   Key
+	value SizedItem
+}
+
+// NewSizedLRU returns a new SizedLRU cache
+func NewSizedLRU(maxSize int64) *SizedLRU {
+	return &SizedLRU{
+		maxSize: maxSize,
+		ll:      list.New(),
+		cache:   make(map[interface{}]*list.Element),
+	}
+}
+
+// Add adds a (key, value) to the cache, evicting items as necessary. Add returns false (
+// and does not add the item) if the item size is larger than the maximum size of the cache.
+func (c *SizedLRU) Add(key Key, value SizedItem) (ok bool) {
```

Comment:
I am thinking ahead when using this actually implement eviction. We would need to return a slice of SizedItem that needs to be deleted from the disk?

Reply:
a callback?